File: test/null.c
#include <unistd.h>
#include <getopt.h>
#include <time.h>

#include <minos.h>
#include "utils.h"


int main(int argc, char** argv)
{
	uint64_t         i;
	mcl_handle**     hdls;
	struct timespec  start, end;
	int              ret = 0;
	
	mcl_banner("NULL Request Test");

	parse_global_opts(argc, argv);

	hdls = (mcl_handle**) malloc(rep * sizeof(mcl_handle*));
	if(!hdls){
		printf("Error allocating memory for MCL handles. Aborting.");
		goto err;
	}
	mcl_init(workers,0x0);

	printf("Synchronous Test...");
	clock_gettime(CLOCK_MONOTONIC,&start);
	for(i=0; i<rep; i++){
		hdls[i] = mcl_task_create();

		if(!hdls[i]){
			printf("Error creating task %" PRIu64 ".\n",i);
			continue;
		}
		
		if(mcl_null(hdls[i])){
			printf("Error submitting request %" PRIu64 ".\n", i);
			continue;
		}
		
		if(mcl_wait(hdls[i]))
			printf("Request %" PRIu64 " timed out!",i);
	}
		
	clock_gettime(CLOCK_MONOTONIC,&end);
	printf("Done. Execution time: %f seconds\n",((float)tdiff(end,start))/BILLION);

	for(i=0; i<rep; i++)
		if(hdls[i]->status != MCL_REQ_COMPLETED){
			printf("Request %"PRIu64" status=%"PRIx64"\n.",
			       i, hdls[i]->status);
			ret = 1;
		}
	
	for(i=0; i<rep; i++)
		mcl_hdl_free(hdls[i]);

	printf("Asynchronous Test...");
	clock_gettime(CLOCK_MONOTONIC,&start);
	for(i=0; i<rep; i++){
		hdls[i] = mcl_task_create();
		if(!hdls[i]){
			printf("Error creating task %" PRIu64 ".",i);
			continue;
		}

		if(mcl_null(hdls[i])){
			printf("Error submitting request %" PRIu64 ".\n", i);
			continue;
		}		
	}
	mcl_wait_all();
	clock_gettime(CLOCK_MONOTONIC,&end);
	printf("Done. Execution time: %f seconds\n",((float)tdiff(end,start))/BILLION);

	for(i=0; i<rep; i++)
		if(hdls[i]->status != MCL_REQ_COMPLETED){
			printf("Request %"PRIu64" status=%"PRIx64"\n.",
			       i, hdls[i]->status);
			ret = 1;
		}

	for(i=0; i<rep; i++)
		mcl_hdl_free(hdls[i]);
	
	mcl_finit();
	mcl_verify(ret);

	free(hdls);
	
	return 0;
 err:
	exit(EXIT_FAILURE);
}

File: rust/libmcl-sys/tests/null.rs
use libmcl_sys::*;

#[test]
fn null_hdl() {
    unsafe {
        let nreqs: usize = 10;
        let mut ret = 0;
        let mut hdls: Vec<*mut mcl_handle> = Vec::<*mut mcl_handle>::new();

        mcl_init(1, MCL_NULL.into());

        println!("Synchronous Test...");

        for i in 0..nreqs {
            hdls.push(mcl_task_create());

            if mcl_null(hdls[i]) == 1 {
                println!("Error submitting request {} ", i);
                continue;
            }

            if mcl_wait(hdls[i]) != 0 {
                println!("Request {} timed out!", i);
            }
        }

        for i in 0..nreqs {
            if (*hdls[i]).status != MCL_REQ_COMPLETED.into() {
                println!("Request {} status={}", i, (*hdls[i]).status);
                ret = 1;
            }
        }

        for i in 0..nreqs {
            mcl_hdl_free(hdls[i]);
        }

        hdls.clear();

        println!("Asynchronous Test...");

        for i in 0..nreqs {
            hdls.push(mcl_task_create());

            if mcl_null(hdls[i]) == 1 {
                println!("Error submitting request {} ", i);
                continue;
            }
        }
        mcl_wait_all();

        for i in 0..nreqs {
            if (*hdls[i]).status != MCL_REQ_COMPLETED.into() {
                println!("Request {} status={}", i, (*hdls[i]).status);
                ret = 1;
            }
        }

        for i in 0..nreqs {
            mcl_hdl_free(hdls[i]);
        }

        mcl_finit();
        assert_eq!(ret, 0);
    }
}
